### DISCUSSION TITLE
packBytes Discussion

### DISCUSSION THREAD

**Discussion by e3dio** [e3dio/packBytes#3] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by jcbbb** [comment]
Hey, do you use this library with uWebsockets.js?

I am using it the following way instead of using json for communication between servers;

```js
// Node.js server
import WebSocket from 'websocket';
import { PackBytes, string } from 'packBytes';

const encoder = new PackBytes({ op: string, payload: string });

const ws = WebSocket('ws://localhost:3020');

ws.on('open', () => {
  const msg = encoder.encode({ op: "subscribe", payload: "some-topic" });
  ws.send(msg);
});

// uWebsockets.js server
import { App } from 'uWebsockets.js';

const app = App({ }).ws('/*', {
  open: (ws) => {
    const ping = encoder.encode({ op: "ping", payload: "connected" })
    ws.send(ping);
  },
  message: (ws, message) => {
    const decoded = encoder.decode(Buffer.from(message));
    console.log(decoded); // { op: "subscribe", payload: "some-topic" }
  }
})
```

What do you think about this? Any recommendations to improve this would be greatly appreciated.

Thanks for this library, I love it.

**Reply by e3dio** [reply]
Hi yeah I am using it for a new project. Normally I would manually code everything but this makes it much easier. You are only using String type, you could convert those to Integer or at least the Op field and it would be much smaller. Here is schema example I am using, your Op field is equivalent to my Type field, I need to add Bits type to the docs:

```javascript
const encoder = new PackBytes({
    Type: bits(8),
    Results: array({
        Name: string,
        Street: string,
        Uuid: blob(16), // 128 bits
        Key: bits(32), // 4,294,967,295
        Price: bits(30), // 1,073,741,823
        Timestamp: bits(22), // 4,194,303
        Lat: bits(20), // 1,048,575
        Lon: bits(20), // 1,048,575
        Area: bits(19), // 524,287
        Zip: bits(11), // 2,047
        City: bits(11), // 2,047
        Year: bits(8), // 255
        Media: bits(8), // 255
        B1: bits(7), // 127
        B2: bits(7), // 127
        Status: bits(4), // 15
    })
});
```

**Reply by e3dio** [reply]
Also you need to send binary messages from uWS.js with `ws.send(buf, true)` you are missing 2nd param

**Reply by jcbbb** [reply]
Let say I have the following schema:

```js
const schema = { op: bits(8), payload: string } // { op: "subscribe", payload: "topic-name" }
```
and a slightly different schema:

```js
const another_schema = { op: bits(8), payload: { topic: bits(12), message: string } } 
// { op: "publish",  payload: { topic: "some-topic", message: "hello-world" } }
```

Is there any way to make payload field kind of universal, or should I create encoders with different schemas?
Or maybe there is a better way to handle this?

What do you think about the following or will it be too large?
```js
const encoder = new PackBytes(array(string))
const buf = encoder.encode(["subscribe", "topic-name"]);
const buf2 = encoder.encode(["publish", "topic-name", "message"]);
```

Thank you for the response.

**Reply by e3dio** [reply]
This is a common scenario that also applies to my case, where you have multiple different schemas that can apply to a message, depending on the type of message. Currently you define each schema individually, then read the 1st byte of the message to determine what schema to use, like how I am doing now:
```javascript

ws.onmessage = (e) => {
   switch (new Uint8Array(e.data, 0, 1)[0]) {
      case 0: schema1.decode(e.data); break;
      case 1: schema2.decode(e.data); break;
      case 2: schema3.decode(e.data); break;
   }
};
```
But since this is common I was thinking about the best way to automate this into the schema system, so you can define 1 schema that includes multiple schemas and it will handle it for you, something like this:
```javascript
const schema = schemas({
   name: schema,
   name2: schema2,
});
```
I will think about the best user friendly way to add that in, for now you can use the 1st method above

**Reply by jcbbb** [reply]
Yeah second option looks great. For now, I will be using first option then

**Reply by e3dio** [reply]
I am thinking about doing it like this, your specific example would be like below with 2 schemas defined, 1 named `subscribe` and 1 named `publish`:
```javascript
const encoder = new PackBytes(schemas({
   subscribe: string,
   publish: { topic: string, message: string }
}));
```
To encode tell it which schema to use and give it the data for that schema:
```javascript
buf1 = encoder.encode('subscribe', 'myTopic');
buf2 = encoder.encode('publish', { topic: 'myTopic', message: 'message123' });
```
Decode tells you which schema was used and value of data:
```javascript
const [ op, data ] = encoder.decode(buf1);
// op == 'subscribe', data == 'myTopic'

const [ op, data ] = encoder.decode(buf2);
// op == 'publish', data == { topic: 'myTopic', message: 'message123' }
```
This would be useful to me so I think I will add it in. This will produce the smallest possible size encoding and allow an easy way to use different named schema types. Let me know if you had any questions, suggestions, or want more schema examples for different data types. I also have a couple more features for strings and arrays coming, like auto-mapping strings to integers, etc.

**Reply by jcbbb** [reply]
Yeah, this would be great to have. I had one question regarding how you share your encoders between servers.
For example, I have an http server and websockets server. Do you define your encoders and put them somewhere like npm and install in projects you need them? I need to use the same encoder in three places (frontend, backend server, websockets server). Do you have any advice on how you would structure these so that encoders are consistent everywhere. Right now, I am just copying encoders between projects :)

**Reply by jcbbb** [reply]
If you used integers instead of strings as keys in `schemas`, would it produce smaller size?
Like this:

```js
const PUBLISH = 1;
const SUBSCRIBE = 2;

const encoder = new PackBytes(schemas({
  [PUBLISH]: { topicName: string, message: string },
  [SUBSCRIBE]: string
}));

const buf = encoder.encode(PUBLISH, { topicName: "some-topic", message: "hello-world" });
const buf2 = encoder.encode(SUBSCRIBE, "some-topic");
```

**Reply by e3dio** [reply]
> If you used integers instead of strings as keys in schemas, would it produce smaller size?

It does exactly that internally, maps the schema name to an integer and only uses the bits required, this example only needs 1 bit to store schema type, so it does it automatically

**Reply by e3dio** [reply]
> how you share your encoders between servers

You have different options for this, manually upload files, automatically fetch new files, you can also serialize the schema into a JSON string and send that over say websocket message, I need to update 1 last thing for that feature tho

**Reply by jcbbb** [reply]
> > how you share your encoders between servers
> 
> You have different options for this, manually upload files, automatically fetch new files, you can also serialize the schema into a JSON string and send that over say websocket message, I need to update 1 last thing for that feature tho

You mean like when websocket connection is established, send over the encoders to the client?

**Reply by e3dio** [reply]
Ya so it is pretty flexible, I haven't actually tried that yet but I built it with that in mind, need to make 1 small update for it to work

**Reply by e3dio** [reply]
Basically just do JSON.stringify(schema) and send that over

**Reply by jcbbb** [reply]
> Ya so it is pretty flexible, I haven't actually tried that yet but I built it with that in mind, need to make 1 small update for it to work

Thanks, I will try it out when it's available then.

**Reply by e3dio** [reply]
I refactored the code and added new features:
- new `schemas` type as discussed above
- JSON serialize any schema with `JSON.stringify(schema)` and open with `new PackBytes(json)`
- `string` added auto-map to integers feature `string('str1', 'str2')` 
- `array` added `size()` method to specify exact size if known

I ran it through the tests and is working well, I will be testing more in production, let me know if you need anything else

**Reply by e3dio** [reply]
I added API section to the README so it is clear how everything works and all options: https://github.com/e3dio/packBytes#api

**Reply by jcbbb** [reply]
Hey, thank you very much for adding these. I tested yesterday and today and it's working great. One question I have now is how do you send just an `op` without any data? For example, I want to ask my server for a new QR-code every 2 minutes, and I don't need to send any data except the `op`. I tried the following:
```js
const encoder = new PackBytes(schemas({
  auth_init: null, // I don't know what to put here to just send the `op`
  auth_pending: string,
  auth_scan: { topic: string, name: string, profile_photo: string } 
}))
```

**Reply by e3dio** [reply]
Makes sense in that case to allow a `null` schema, I added that in latest commit, should work for you now as expected. Good feature, let me know if you need anything else

**Reply by jcbbb** [reply]
Thank you, it's working now :)

**Comment by jcbbb** [comment]
Hey, I was trying to encode object that has the following shape:

```javascript
{
  createdBy: 12,
  id: 2,
  posting: {
    id: 2,
    title: "Hello world"
  },
  members: [{name: "a", id: 2}]
}
```
and the schema I used is the following:

```javascript
let schema = schemas({
  new_chat: {
    createdBy: bits(32),
    id: bits(32),
    posting: {
      id: bits(32),
      title: string    
    },
    members: array({name: string, id: bits(32)})
  }
})
```

But this gives me varInt max exceeded error.

I tried to wrap that nested posting object in new PackBytes like this

```javascript
let schema = schemas({
  new_chat: {
    // other stuff
    posting: new PackBytes({
       id: bits(32),
       title: string
    })
  }
})
```

Getting same error as above. Do you know how to write schema for objects like this? :) Thank you

**Reply by e3dio** [reply]
Your code works fine for me, the only thing missing in your post is `encoder.encode('new_chat', data)` here is in full:

```js
import { bool, bits, string, array, float, blob, schemas, PackBytes } from 'packbytes';

const data = {
  createdBy: 12,
  id: 2,
  posting: {
    id: 2,
    title: "Hello world"
  },
  members: [{name: "a", id: 2}]
}

const schema = schemas({
  new_chat: {
    createdBy: bits(32),
    id: bits(32),
    posting: {
      id: bits(32),
      title: string    
    },
    members: array({name: string, id: bits(32)})
  }
})

const encoder = new PackBytes(schema);

const buf = encoder.encode('new_chat', data);

console.log(buf);

const [ type, data ] = encoder.decode(buf);

console.log(type, data);
```

**Reply by jcbbb** [reply]
Oh shit, forgot I to pass that `new_chat` op as first argument :(
This whole time I have just been passing only the data. Sorry to bother you with this nonsense
